### DISCUSSION TITLE
void를 반환하는 메서드는 어떻게 테스트 하는게 좋은 방법일까요?

### DISCUSSION THREAD

**Discussion by Kwonminwoo** [SCBJ-7/SCBJ-BE#63] [Q&A]
### 발생과정
```java
@Transactional
public void deleteProduct(Long productId){
    Product targetProduct = productRepository.findById(productId)
        .orElseThrow(() -> new ProductNotFoundException(ErrorCode.PRODUCT_NOT_FOUND));

    targetProduct.delete(LocalDateTime.now());
}
```
상품 삭제의 서비스 메서드는 `void`를 반환합니다.

### 원인

- 삭제 메서드는 `void`를 반환하기 때문에 기존의 `given().willReturn()` 방식으로는 코드를 작성할 수 없습니다. (`willReturn()`에 값을 못넣음)

### 해결방안
`controller` 테스트 코드는
```java
// given
Long targetProductId = 1L;

doNothing().when(productService).deleteProduct(any());
```
`Mockito`의 `doNothing()`을 사용합니다. `doNoting()`은 메서드의 실행 시키지 않고 아무것도 하지 않도록 하는 것입니다. 
어차피 void를 반환하기 때문에 `controller` 테스트에서는 해당 메서드가 어떤 로직을 가지고 있던지 영향을 끼치지 않기 때문에, `doNothing()`을 사용하는것도 문제가 없을 것 같습니다. 

다음은 `service` 테스트 코드 입니다.
```java
given(productRepository.findById(any())).willReturn(Optional.of(product));

// when
productService.deleteProduct(1L);

// then
verify(productRepository, atLeastOnce()).findById(any());
```
`deleteProduct() `메서드가 `void`를 반환하므로 반환값을 통해 검증할 수가 없습니다. 그래서 `verify()`를 이용해 해당 메서드가 실행 되었는지를 검증할 수 있습니다.

혹시 다른 방법이 있으시면 추천 부탁드려요!!~!!

**Comment by YurimYang** [comment]
저도 update시에 이문제를 어떻게 해결할까 하다가 verify로 실행 검증만 진행했습니다 ! 이전 프로젝트에서 다들 그렇게 진행하시더라고요 ㅎㅎ
